### PR TITLE
NavigationBuilder AddAsync().

### PR DESCRIPTION
**Pull request by jtkech**
To prevent, in some places, from assigning an async void lambda to an action parameter.

### CODE REVIEW

**Comment by sebastienros** [comment]
As long as the `itemBuilder` itslef is awaited, I don't see a reason to make this one async.

**Comment by jtkech** [comment]
Not sure but just because here,

we were passing an async void lambda to a `Action<T>` parameter, not a `Func<T, Task>`.

    builder.Add(new LocalizedString(node.LinkText, node.LinkText), async itemBuilder => {

**Comment by sebastienros** [comment]
So the issue is the same, each of these lambda should be awaited. Make the builder only accept `Func<T, Task>` and ensure their are awaited.

**Comment by jtkech** [comment]
But `builder.Add()` is used in many place without async lambda and no await.

Do you mean that in these case i would have to add `return Task.CompletedTask`?

If so, do i still have to rename `Add()` to `AddAsync()`?

**Comment by sebastienros** [comment]
The Add is not Async, the Add can add a `Func<T,Task>` but when this lambda is used, it should be awaited. It can take an override that accepts an Action<T> but in this case it's just a wrapper around a Func<T,Task> that returns Task.Completed.

The issue is really that when the lambda is accessed then there it should be awaited. And I haven't seend this change yet.

**Comment by jtkech** [comment]
The added line in `NavigationBuilder.cs` line 44.

**Comment by sebastienros** [comment]
ok didn't see that. Create an Add and an AddAsync then. Add calls into AddAsync by wrapping the Action to a Func<T,Task> returning a Task.Completed.

**Comment by jtkech** [comment]
Okay

File: src/OrchardCore/OrchardCore.Navigation.Core/NavigationBuilder.cs [hunk 1]
```diff
@@ -14,25 +15,42 @@ public NavigationBuilder()
...
 
             return this;
         }
 
+        public Task<NavigationBuilder> AddAsync(LocalizedString caption, Func<NavigationItemBuilder, Task> itemBuilder, IEnumerable<string> classes = null)
+        {
+            return AddAsync(caption, null, itemBuilder, classes);
+        }
+
+        public NavigationBuilder Add(LocalizedString caption, string position, Action<NavigationItemBuilder> itemBuilder, IEnumerable<string> classes = null, int priority = 0)
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.Navigation.Core/NavigationBuilder.cs]: Yeah, sorry, I didn't see that the itembuilder was called in the body. I thought it was called afterward :/

[the remark this answers is not in the document]
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Navigation.Core/NavigationBuilder.cs]: No problem ;)

**Comment by sebastienros** [comment]
How did you find the issue?

**Comment by jtkech** [comment]
As i remember after a comment on gitter saying that the admin menu of the demo module was not rendered. So, i took a quick look in another module to see how an admin menu is now implemented.

Then, the first one i saw was using an async lambda, and now it's automatic for me, in this case i always check the parameter type ;)

**Comment by sebastienros** [comment]
ok, merge it when ready

**Pull request merged**